### ISSUE TITLE
Support reproducible builds 

### ISSUE THREAD

**Issue by salrashid123** [replicate/cog#1250]
Cog currently [uses docker to build the images](https://github.com/replicate/cog/blob/main/pkg/docker/build.go#L42)

however, docker based builds are not  reproducible: you'll get different image hashes even with the identical config

this long-term feature request is to refactor the build system from docker to something like 

* [kaniko](https://github.com/GoogleContainerTools/kaniko#flag---reproducible)
* `buildah`
* or at absolute worst case, `bazel`
   * [Building deterministic Docker images with Bazel](https://blog.bazel.build/2015/07/28/docker_build.html#building-deterministic-docker-images-with-bazel)

---

some references  building using kaniko and bazel

* [cosign_kaniko_cloud_build](https://github.com/salrashid123/cosign_kaniko_cloud_build)
* [cosign_bazel_cloud_build](https://github.com/salrashid123/cosign_bazel_cloud_build)

**Comment by technillogue** [comment]
Hi, we've investigated this - SOURCE_DATE_EPOCH is a promising direction, and we [tried approaches](https://github.com/technillogue/timestop-rs) with resetting mtime for everything. Unfortunately, `pip install` is fundamentally irreproducible, because it generates pyc files that include the timestamp. Unzipping wheels without using pip might make this possible, or I think there's some PEPs in the works that might help with this.

**Comment by salrashid123** [comment]
got it; i think esp with python it'd be difficult to do with its own toolchains.

maybe generating the docker file  per https://github.com/replicate/cog/issues/1241#issuecomment-1660128528

and then chaining it to off the shelf `kaniko` would be sufficient workaround ( [ref](https://github.com/salrashid123/tpm_daemonset/tree/main#build)

```
docker run   \
  -v `pwd`:/workspace -v $HOME/.docker/config_docker.json:/kaniko/.docker/config.json:ro \
   -v /var/run/docker.sock:/var/run/docker.sock \
     gcr.io/kaniko-project/executor@sha256:034f15e6fe235490e64a4173d02d0a41f61382450c314fffed9b8ca96dff66b2    \
	 --dockerfile=Dockerfile \
	 --reproducible   \
	     --destination "docker.io/salrashid123/tpmds:server"       --context dir:///workspace/
```

i realize now we're involving kaniko as well but it maybe easier to delegate it like this for now

**Comment by technillogue** [comment]
Would that address the pyc timestamps?

**Comment by salrashid123** [comment]
i think so, as part of the kaniko reproducible builds, it sets up snapshots resetting the all file times. 

tried it from the [getting started](https://github.com/replicate/cog/blob/main/docs/getting-started.md) guide and using the generated `Dockerfile` seems to always reference a file like 

```
COPY .cog/tmp/build1866459875/cog-0.0.1.dev-py3-none-any.whl /tmp/cog-0.0.1.dev-py3-none-any.whl
```

which doesn't exist and causes the kanilo to fail

```
cog build
cog debug  > Dockerfile

docker run     -v `pwd`:/workspace -v $HOME/.docker/config_docker.json:/kaniko/.docker/config.json:ro    -v /var/run/docker.sock:/var/run/docker.sock      gcr.io/kaniko-project/executor@sha256:034f15e6fe235490e64a4173d02d0a41f61382450c314fffed9b8ca96dff66b2    	 --dockerfile=Dockerfile 	 --reproducible   	     --destination "docker.io/salrashid123/cogdemo:server"       --context dir:///workspace/
```

---

**Comment by salrashid123** [comment]
oh, so python embeds the timestamp _inside_ the file...then kaniko isn't gonna help out.

...and i can't sincerely recommend going all out on and investing in [python-bazel builds](https://github.com/salrashid123/python-bazel-docker)

**Comment by technillogue** [comment]
as a stopgap for the debug issue, run cog build once and then interrupt, it will place a cog wheel in .cog/tmp/whatever, and then you can edit the `cog debug` output

does python-bazel address pyc timestamps somehow? does it just strip pyc files? 

it would be incredibly helpful for us to get reproducible builds for deduplication

**Comment by salrashid123** [comment]
yeah, i tried the interrupt trick suggested but each cog+kaniko build is different hash (which is expected, i tihnk)

i'm unsure exactly how bazel `rules_python` handles pyc files but i can say you need to precisely  define everything upfront and bazel uses its own sandbox to canonicalize everything.   

some examples with rules_python which may help answer the question though....once it works with rules_python, stitching it with rules_docker and containers would be easy   

https://github.com/bazelbuild/rules_python/tree/main/examples

**Comment by charles-dyfis-net** [comment]
Y'all might also investigate Nix (which provides `dockerTools`, an alternate build toolchain for Docker images) towards this end.

**Comment by technillogue** [comment]
Does rules_python generate pyc at all? https://github.com/bazelbuild/bazel/issues/16245

Again, there's no issue with mtimes, the problem is the timestamps embedded in pyc files

**Comment by charles-dyfis-net** [comment]
> Does rules_python generate pyc at all? [bazelbuild/bazel#16245](https://github.com/bazelbuild/bazel/issues/16245)
> 
> Again, there's no issue with mtimes, the problem is the timestamps embedded in pyc files

The NixOS install CD is fully binary reproducible. I can't imagine it not including Python, so clearly they've got that licked somehow.

**Comment by technillogue** [comment]
then we would have to ship nix's patched interpreter, right? DETERMINISTIC_BUILD is not present in stock python

**Comment by salrashid123** [comment]
her'es an end-to-end covering building an image with [bazel and serving with cog](https://gist.github.com/salrashid123/1509254ac4cef16dcc8e9e2f257599cd).

if precise build steps are followed, you should end up with 

* `sha256:3db6542dc746aeabaa39d902570430e1d50c416e7fc20b875c10578aa5e62875`

(i verified it on two different clean vms)

as mentioned, using bazel is really tedious though toolchains like gazelle _may_ help with python.
(imo as-is in current state, the developer friction all this introduces negates the primary ease-of-use benefits of using/building w/  cog in the first place)

[tbh, i've never used or needed cog and try to not use bazel for deterministic builds (in go there are easier ways)...this issue with cog was something i noticed and then ratholed academically.]

**Comment by RyzeNGrind** [comment]
I would like to add my +1 for supporting reproducible builds via Nix and NixOS as well.

**Comment by technillogue** [comment]
https://github.com/datakami/cognix is a project that exists and kind of works but unfortunately isn't a priority for us at this time

**Comment by markphelps** [comment]
Closing this as stale. After thorough investigation, reproducible builds are blocked by fundamental limitations in the Python toolchain (pip-generated pyc timestamps), and viable workarounds like Bazel or Nix introduce significant complexity that conflicts with Cog’s ease-of-use goals. The team confirmed this is not on the current roadmap, and there has been no activity for over two years. If the Python ecosystem changes or a low-friction solution emerges, we’re happy to revisit. Thanks everyone for the detailed exploration.
